apps/emqx_connector/test/emqx_connector_mysql_SUITE.erl

```diff
@@ -0,0 +1,143 @@
+% %%--------------------------------------------------------------------
+% %% Copyright (c) 2020-2022 EMQ Technologies Co., Ltd. All Rights Reserved.
+% %%
+% %% Licensed under the Apache License, Version 2.0 (the "License");
+% %% you may not use this file except in compliance with the License.
+% %% You may obtain a copy of the License at
+% %% http://www.apache.org/licenses/LICENSE-2.0
+% %%
+% %% Unless required by applicable law or agreed to in writing, software
+% %% distributed under the License is distributed on an "AS IS" BASIS,
+% %% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+% %% See the License for the specific language governing permissions and
+% %% limitations under the License.
+% %%--------------------------------------------------------------------
+
+-module(emqx_connector_mysql_SUITE).
+
+-compile(nowarn_export_all).
+-compile(export_all).
+
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+-define(MYSQL_HOST, "mysql").
+-define(MYSQL_PORT, 3306).
+
+all() ->
+    emqx_common_test_helpers:all(?MODULE).
+
+groups() ->
+    [].
+
+init_per_suite(Config) ->
+    case emqx_common_test_helpers:is_tcp_server_available(?MYSQL_HOST, ?MYSQL_PORT) of
+        true ->
+            Config;
+        false ->
+            {skip, no_mysql}
+    end.
+
+end_per_suite(_Config) ->
+    ok.
+
+init_per_testcase(_, Config) ->
+    ?assertEqual(
+        {ok, #{poolname => emqx_connector_mysql}},
+        emqx_connector_mysql:on_start(<<"emqx_connector_mysql">>, mysql_config())
+    ),
+    Config.
+
+end_per_testcase(_, _Config) ->
+    ?assertEqual(
+        ok,
+        emqx_connector_mysql:on_stop(<<"emqx_connector_mysql">>, #{poolname => emqx_connector_mysql})
+    ).
+
+% %%------------------------------------------------------------------------------
+% %% Testcases
+% %%------------------------------------------------------------------------------
+
+% Simple test to make sure the proper reference to the module is returned.
+t_roots(_Config) ->
+    ExpectedRoots = [{config, #{type => {ref, emqx_connector_mysql, config}}}],
+    ActualRoots = emqx_connector_mysql:roots(),
+    ?assertEqual(ExpectedRoots, ActualRoots).
+
+% Not sure if this level of testing is appropriate for this function.
+% Checking the actual values/types of the returned term starts getting
+% into checking the emqx_connector_schema_lib.erl returns and the shape
+% of expected data elsewhere.
```

Comment:
Maybe this seems like testing Hocon itself?

Reply:
It's not Hocon. The returns from the function `emqx_connector_mysql:fields(config),` are hardcoded. For example:
```
relational_db_fields() ->
    [ {server, fun server/1}
    , {database, fun database/1}
    , {pool_size, fun pool_size/1}
    , {username, fun username/1}
    , {password, fun password/1}
    , {auto_reconnect, fun auto_reconnect/1}
    ].
```

The only reason I put the check in there the way it is, is that technically you don't have to call the `emqx_connector_mysql` module to return these values that are expected. You could hardcode them yourselves such as in the redis callback modue:
```
fields(single) ->
    [ {server, #{type => server()}}
    , {redis_type, #{type => hoconsc:enum([single]),
                     default => single}}
    ] ++
    redis_fields() ++
    emqx_connector_schema_lib:ssl_fields();
```

So there is an expected shape of the data that needs to be returned, but how that data is actually defined is entirely up in the air. What I'm checking for appears to be the minimum viable/expected shape of the data.

Reply:
I see, that makes sense.

Then, maybe if we added some dialyzer specs to those functions, it might help keep track of correct types. :smile_cat: